### PR TITLE
directory service

### PR DESCRIPTION
**Pull request by besseym**
Created a directory service to interact with Active Directory

### CODE REVIEW

File: mmvizutil/directory/ldap.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        finally:
+            if connection:
+                connection.unbind()
+
+        return result_list
+
+    def get_record(self, mm_id, base = "dc=na,dc=mmfg,dc=net"):
+
+        result_list = self.search(base, f"(samaccountname={mm_id})")
+        dn, record = result_list[0]
```
Comment (dmitrypolo) [inline mmvizutil/directory/ldap.py]: looks like we aren't using `dn` anywhere else so no need to keep it, can just do `_, record = result_list[0]`

File: mmvizutil/directory/ldap.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        ldap.set_option(ldap.OPT_X_TLS_REQUIRE_CERT, ldap.OPT_X_TLS_NEVER)
+
+    def search(self, base, query):
+
+        connection = None
+        result_list = None
+
+        try:
+
+            connection = ldap.initialize(self.host)
```
Comment (dmitrypolo) [inline mmvizutil/directory/ldap.py]: should we decouple the connection from the search function? if we want to perform multiple searches each search initializes a new connection, I am not sure if this is expensive or not?

File: mmvizutil/directory/tests.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    LdapService
+)
+
+class TestLdapService(unittest.TestCase):
+
+    def setUp(self):
+
+        config = configparser.ConfigParser()
+        config_path = os.path.join(os.path.abspath('..'), "config/ldap.ini")
+        print(config_path)
```
Comment (dmitrypolo) [inline mmvizutil/directory/tests.py]: probably don't need print statements in the setup

File: mmvizutil/directory/tests.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        config.read(config_path)
+        print(config.sections())
+
+        config_ldap = config["LDAP-PROD"]
+
+        host = config_ldap.get("HOST")
+        username = config_ldap.get("USERNAME")
+        password = config_ldap.get("PASSWORD")
+
+        self.ldap_service = LdapService(host, username, password)
```
Comment (dmitrypolo) [inline mmvizutil/directory/tests.py]: shouldn't we be making mock calls to the `ldap` server?

File: mmvizutil/directory/tests.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    def test_get_person(self):
+
+        record = self.ldap_service.get_record("mm28969", "dc=na,dc=mmfg,dc=net")
+        person = LdapPerson()
+        person.parse(record)
+
+        self.assertDictEqual(person.__dict__, self.person_dict)
+        print(person.__dict__)
+
+
```
Comment (dmitrypolo) [inline mmvizutil/directory/tests.py]: going back to the decoupling of connection and searching in the tests alone we are making 3 separate calls to the server, can get expensive as tests pile up and as this class gets used

File: mmvizutil/directory/ldap.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        result_list = self.search(base, f"(samaccountname={mm_id})")
+        dn, record = result_list[0]
+
+        return record
+
+    def get_groups(self, mm_id, base = "dc=na,dc=mmfg,dc=net"):
+
+        record = self.get_record(mm_id, base)
+
+        member_of = record['memberOf']
```
Comment (dmitrypolo) [inline mmvizutil/directory/ldap.py]: is it possible for this to be empty? I am not familiar with `ldap` enough to know.

File: mmvizutil/directory/ldap.py (hunk 2 above)
Comment (besseym) [reply mmvizutil/directory/ldap.py]: I agree, good to decouple the connection

File: mmvizutil/directory/tests.py (hunk 4 above)
Comment (besseym) [reply mmvizutil/directory/tests.py]: Yes, I would like to discuss how we would do this.

File: mmvizutil/directory/ldap.py (hunk 6 above)
Comment (besseym) [reply mmvizutil/directory/ldap.py]: not sure, better to be safe

**Comment by besseym** [comment]
Don't worry about making mock tests yet. I'm going to make a plan for everyone to address mock tests for mmvizutil and metadata_service. We will need your expertise then.

**Pull request merged**